### PR TITLE
Build module - basic resource payment

### PR DESCRIPTION
**Pull request by veteran29**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes|
| Needs wipe? | no |
| Fixed issues | #575 (partially) |

### Description:

Very basic implementation of resource payment in build module. To be improved as some sort of build queue in future.

Additional fixes done by the way.

### Content:
- [x] Resource payment for when building items (server side), simple implementation to be extended in future
- [x] Search filter in build menu
- [x] Fixed bug with undeletable local objects when switching between build/move modes
- [x] Auto enable functions recompile in 3DEN multiplayer
- [x] Base clases right panel controls

### Successfully tested on:
- [x] Local MP
- [x] Dedicated MP

### Compatibility checked with:
* NOTHING

### CODE REVIEW

**Comment by veteran29** [comment]
@KillahPotatoes/kplib-testers open for testing. 

Also needs testing on dedi, I'm out of home till end of week. @KillahPotatoes/kp-liberation

File: Missionframework/description.ext [hunk 1]
```diff
@@ -151,7 +151,7 @@ allowFunctionsLog = 0;
 
 // As a security measure, functions are by default protected against rewriting during mission.
 // This restriction does not apply in missions previewed from the editor.
-allowFunctionsRecompile = 0;
+allowFunctionsRecompile = __EVAL([0, 1] select (is3DENMultiplayer || is3DEN));
```
Comment (Wyqer) [inline Missionframework/description.ext]: What about the comment above concerning not apply in mission preview from the editor? Wrong hint or just overlooked?

File: Missionframework/modules/15_build/ui/buildDisplay_left.hpp [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+class KPLIB_BuildSearch: KPGUI_PRE_EditBox {
+    idc = KPLIB_IDC_BUILD_SEARCH;
+
+    x = KP_GETCX(KP_X_VAL_LP,KP_WIDTH_VAL_LP,0,12);
+    y = KP_GETCY(KP_Y_VAL_LP,KP_HEIGHT_VAL_LP,3,40);
+    w = KP_GETW(KP_WIDTH_VAL_LP,(12/11));
+    h = KP_GETH(KP_HEIGHT_VAL_LP,40);
+
+    onKeyUp = "call KPLIB_fnc_build_displayFillList";
```
Comment (Wyqer) [inline Missionframework/modules/15_build/ui/buildDisplay_left.hpp]:
```suggestion
    onKeyUp = "[] call KPLIB_fnc_build_displayFillList";
```
Concerning coding convention, no calls/spawns without parameters (or empty parameters)

File: Missionframework/modules/15_build/ui/buildDisplay_left.hpp [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+class KPLIB_BuildSearchClearButton: KPGUI_PRE_ActivePictureRatio {
+    idc = KPLIB_IDC_BUILD_SEARCH_BUTTON;
+
+    x = KP_GETCX(KP_X_VAL_LP,KP_WIDTH_VAL_LP,11,12);
+    y = KP_GETCY(KP_Y_VAL_LP,KP_HEIGHT_VAL_LP,3,40);
+    w = KP_GETW(KP_WIDTH_VAL_LP,(12/1));
+    h = KP_GETH(KP_HEIGHT_VAL_LP,40);
+
+    text = "\a3\Ui_f\data\GUI\RscCommon\RscButtonSearch\search_end_ca.paa";
+    action = "call KPLIB_fnc_build_searchClear";
```
Comment (Wyqer) [inline Missionframework/modules/15_build/ui/buildDisplay_left.hpp]:
```suggestion
    action = "[] call KPLIB_fnc_build_searchClear";
```

File: Missionframework/modules/15_build/ui/buildDisplay_left.hpp [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+class KP_ApplyButton: KPGUI_PRE_DialogButton_LeftPanel {
+    idc = KPLIB_IDC_BUILD_CONFIRM;
+    text = "$STR_KPLIB_DIALOG_BUTTON_BUILD";
+
+    w = KP_GETWPLAIN(KP_WIDTH_VAL_LP,1);
+    onButtonClick = "_this call KPLIB_fnc_build_confirmAll";
+};
+
+class KPLIB_DialogCross: KPGUI_PRE_DialogCross_LeftPanel {
+    action = "call KPLIB_fnc_build_stop";
```
Comment (Wyqer) [inline Missionframework/modules/15_build/ui/buildDisplay_left.hpp]:
```suggestion
    action = "[] call KPLIB_fnc_build_stop";
```

File: Missionframework/stringtable.xml [hunk 5]
```diff
@@ -574,6 +574,11 @@
             <Turkish>İnşa Et</Turkish>
         </Key>
 
+        <!-- BUILD DIALOG -->
+        <Key ID="STR_KPLIB_DIALOG_BUILD_QUEUE_TITLE">
+            <English>BUILD QUEUE</English>
```
Comment (Wyqer) [inline Missionframework/stringtable.xml]:
```suggestion
            <English>BUILD QUEUE</English>
            <German>BAULISTE</German>
```

File: Missionframework/stringtable.xml [hunk 6]
```diff
@@ -1060,6 +1065,9 @@
             <Russian>Следующие элементы не разрешены:</Russian>
             <Turkish>Şu eşyalar yasaklı:</Turkish>
         </Key>
+        <Key ID="STR_KPLIB_HINT_BUILD_NO_RESOURCES">
+            <English>Not enough resources to build %1.</English>
```
Comment (Wyqer) [inline Missionframework/stringtable.xml]:
```suggestion
            <English>Not enough resources to build %1.</English>
            <German>Nicht genug Ressourcen, um %1 zu bauen.</German>
```

File: Missionframework/KPGUI/KPGUI_defines.hpp [hunk 7]
```diff
@@ -142,6 +142,16 @@
 #define KP_WIDTH_VAL_LP                 0.2
 #define KP_HEIGHT_VAL_LP                (1 - 2 * KP_Y_VAL_LP - KP_HEIGTH_TITLE - KP_HEIGTH_BUTTON - 2 * KP_SPACING_Y)
 
+/*
+    --- Right panel dialog components ---
+*/
+
+#define KP_X_VAL_RP                     0.8 - 0.0025
```
Comment (Wyqer) [inline Missionframework/KPGUI/KPGUI_defines.hpp]: Why not 0.7975?

File: Missionframework/KPGUI/KPGUI_defines.hpp [hunk 8]
```diff
@@ -142,6 +142,16 @@
 #define KP_WIDTH_VAL_LP                 0.2
 #define KP_HEIGHT_VAL_LP                (1 - 2 * KP_Y_VAL_LP - KP_HEIGTH_TITLE - KP_HEIGTH_BUTTON - 2 * KP_SPACING_Y)
 
+/*
+    --- Right panel dialog components ---
+*/
```
Comment (Wyqer) [inline Missionframework/KPGUI/KPGUI_defines.hpp]:
```suggestion
    (X from 0.7975 - 0.9975, Y from 0.0025 - 0.9975)
*/
```
For the left panel dialog:
```
    (X from 0.0025 - 0.2025, Y from 0.0025 - 0.9975)

File: Missionframework/KPGUI/KPGUI_defines.hpp [hunk 9]
```diff
@@ -142,6 +142,16 @@
...
+
+#define KP_X_VAL_RP                     0.8 - 0.0025
+#define KP_Y_VAL_RP                     0.0025
+
+#define KP_WIDTH_VAL_RP                 0.2
+#define KP_HEIGHT_VAL_RP                (1 - 2 * KP_Y_VAL_RP - KP_HEIGTH_TITLE - KP_HEIGTH_BUTTON - 2 * KP_SPACING_Y)
```
Comment (Wyqer) [inline Missionframework/KPGUI/KPGUI_defines.hpp]:
```suggestion
#define KP_HEIGHT_VAL_RP                (0.905 - KP_HEIGTH_TITLE - KP_HEIGTH_BUTTON - 2 * KP_SPACING_Y)
```
same for left panel dialog

File: Missionframework/modules/15_build/ui/buildDisplay_left.hpp [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+
+class KPLIB_BuildList: KPGUI_PRE_ListNBox {
+    idc = KPLIB_IDC_BUILD_ITEM_LIST;
+
+    x = KP_GETCX(KP_X_VAL_LP,KP_WIDTH_VAL_LP,0,1);
+    y = KP_GETCY(KP_Y_VAL_LP,KP_HEIGHT_VAL_LP,2,20);
+    w = KP_GETW(KP_WIDTH_VAL_LP,1);
+    h = KP_GETH(KP_HEIGHT_VAL_LP,(20/18));
+
+    columns[] = { 0, 0.65, 0.75, 0.85 };
```
Comment (Wyqer) [inline Missionframework/modules/15_build/ui/buildDisplay_left.hpp]:
```suggestion
    columns[] = {0, 0.65, 0.75, 0.85};
```

File: Missionframework/modules/15_build/ui/buildDisplay_left.hpp [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+    x = KP_GETCX(KP_X_VAL_LP,KP_WIDTH_VAL_LP,0,1);
+    y = KP_GETCY(KP_Y_VAL_LP,KP_HEIGHT_VAL_LP,0,20);
+    w = KP_GETW(KP_WIDTH_VAL_LP,1);
+    h = KP_GETH(KP_HEIGHT_VAL_LP,20);
+
+    // Toolbox Controls
+    class Controls {
+        // !TODO! move toolbox items creation to script
+        class KPLIB_Toolbox_MoveItems: KPGUI_PRE_ActiveText {
+            text = "Build mode";
```
Comment (Wyqer) [inline Missionframework/modules/15_build/ui/buildDisplay_left.hpp]: Missing stringtable key

File: Missionframework/modules/15_build/ui/buildDisplay_left.hpp [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+            tooltip = "Click to toggle between building new items and moving existing items modes."
```
Comment (Wyqer) [inline Missionframework/modules/15_build/ui/buildDisplay_left.hpp]: Missing stringtable key

File: Missionframework/modules/15_build/fnc/fn_build_displayPlaceObject.sqf [hunk 13]
```diff
@@ -36,6 +36,8 @@ if !(_className isEqualTo "") exitWith {
     _obj setPosASL _cursorWorldPosASL;
     _obj setVectorUp _cursorSurfaceNormal;
 
+    _obj setVariable ["KPLIB_buildPrice", [_priceSupplies, _priceAmmo, _priceFuel]];
```
Comment (Wyqer) [inline Missionframework/modules/15_build/fnc/fn_build_displayPlaceObject.sqf]:
```suggestion
    _obj setVariable ["KPLIB_build_price", [_priceSupplies, _priceAmmo, _priceFuel]];
```

File: Missionframework/modules/15_build/fnc/fn_build_confirmAll.sqf [hunk 14]
```diff
@@ -21,16 +21,16 @@
...
 LSVAR("buildQueue", LGVAR(buildQueue) - _validItems);
 
-// !TODO! implement build queue handling (resource check etc.)
-systemChat "buildConfirm: Resource check not implemented yet!";
 {
     private _dirAndUp = [vectorDir _x, vectorUp _x];
     private _pos = getPosATL _x;
     private _class = typeOf _x;
 
+    private _price = _x getVariable "KPLIB_buildPrice";
```
Comment (Wyqer) [inline Missionframework/modules/15_build/fnc/fn_build_confirmAll.sqf]:
```suggestion
    private _price = _x getVariable "KPLIB_build_price";
```

File: Missionframework/description.ext (hunk 1 above)
Comment (veteran29) [reply Missionframework/description.ext]: In my experience it was working only in SP preview.

File: Missionframework/modules/15_build/ui/buildDisplay_left.hpp (hunk 3 above)
Comment (Dubjunk) [inline Missionframework/modules/15_build/ui/buildDisplay_left.hpp]:
```suggestion

```
Can be removed as the onKeyUp event doesn't block the backspace/delete keyboard button.

File: Missionframework/modules/15_build/fnc/fn_build_searchClear.sqf [hunk 15]
```diff
@@ -0,0 +1,29 @@
+#include "..\ui\defines.hpp"
+#include "script_components.hpp"
+/*
+    KPLIB_fnc_build_searchClear
+
+    File: fn_build_searchClear.sqf
```
Comment (Dubjunk) [inline Missionframework/modules/15_build/fnc/fn_build_searchClear.sqf]: FIle can be deleted, the isn't necessary anymore.

File: Missionframework/modules/15_build/ui/buildDisplay_left.hpp [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+
+    onMouseZChanged = "['onMouseZChanged_BuildCategoryList', _this] call KPLIB_fnc_build_handleMouse";
+};
+
+class KPLIB_BuildSearch: KPGUI_PRE_EditBox {
+    idc = KPLIB_IDC_BUILD_SEARCH;
+
+    x = KP_GETCX(KP_X_VAL_LP,KP_WIDTH_VAL_LP,0,12);
+    y = KP_GETCY(KP_Y_VAL_LP,KP_HEIGHT_VAL_LP,3,40);
+    w = KP_GETW(KP_WIDTH_VAL_LP,(12/11));
```
Comment (Dubjunk) [inline Missionframework/modules/15_build/ui/buildDisplay_left.hpp]:
```suggestion
    w = KP_GETW(KP_WIDTH_VAL_LP,1);
```

File: Missionframework/modules/15_build/ui/buildDisplay_left.hpp [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+    h = KP_GETH(KP_HEIGHT_VAL_LP,40);
+
+    onKeyUp = "call KPLIB_fnc_build_displayFillList";
+};
+
+class KPLIB_BuildSearchClearButton: KPGUI_PRE_ActivePictureRatio {
```
Comment (Dubjunk) [inline Missionframework/modules/15_build/ui/buildDisplay_left.hpp]: Can be deleted.

File: Missionframework/description.ext (hunk 1 above)
Comment (Wyqer) [reply Missionframework/description.ext]: Ok. As it won't hurt in that implementation with __EVAL, it's fine and we can leave it that way. 👍

File: Missionframework/modules/15_build/ui/buildDisplay_left.hpp (hunk 3 above)
Comment (veteran29) [reply Missionframework/modules/15_build/ui/buildDisplay_left.hpp]: I left it as it's convivient to have clear search button IMO.

Comment (veteran29) [reply Missionframework/modules/15_build/ui/buildDisplay_left.hpp]: I left it as it's convivient to have clear search button IMO.

**Pull request merged**